File: src/main/java/com/meti/api/nulls/None.java
package com.meti.api.nulls;

import java.util.function.Consumer;
import java.util.function.Function;
import java.util.function.Supplier;

public class None<T> implements Option<T> {
    private None() {
    }

    public static <T> None<T> None() {
        return new None<>();
    }

    @Override
    public Option<T> peek(Consumer<T> consumer) {
        return None();
    }

    @Override
    public T orElseSupply(Supplier<T> supplier) {
        return supplier.get();
    }

    @Override
    public <R> Option<R> map(Function<T, Option<R>> function) {
        return None();
    }
}

File: src/main/java/com/meti/None.scala
package com.meti

class None[T]() extends Option[T] {
  override def orElse(t: T): T = t

  def apply[R](): None[R] = new None()

  override def map[R](mapper: F1[T, R]): Option[R] = new None[R]
}